sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestStream.java

```diff
@@ -0,0 +1,223 @@
+
+      currentWatermark = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * Adds the specified elements to the source with timestamp equal to the current watermark.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        T element, T... elements) {
+      TimestampedValue<T> firstElement = TimestampedValue.of(element, currentWatermark);
+      @SuppressWarnings("unchecked")
+      TimestampedValue<T>[] remainingElements = new TimestampedValue[elements.length];
+      for (int i = 0; i < elements.length; i++) {
+        remainingElements[i] = TimestampedValue.of(elements[i], currentWatermark);
+      }
+      return addElements(firstElement, remainingElements);
+    }
+
+    /**
+     * Adds the specified elements to the source with the provided timestamps.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      events.add(ElementEvent.add(element, elements));
+      return this;
+    }
+
+    /**
+     * Advance the watermark of this source to the specified instant.
+     *
+     * <p>The watermark must advance monotonically and to at most
+     * {@link BoundedWindow#TIMESTAMP_MAX_VALUE}.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceWatermarkTo(Instant newWatermark) {
+      checkArgument(newWatermark.isAfter(currentWatermark),
+          "The watermark must monotonically advance");
+      checkArgument(newWatermark.isBefore(BoundedWindow.TIMESTAMP_MAX_VALUE),
+          "The Watermark cannot progress beyond the maximum. Got: %s. Maximum: %s",
+          newWatermark,
+          BoundedWindow.TIMESTAMP_MAX_VALUE);
+      events.add(WatermarkEvent.<T>advanceTo(newWatermark));
+      currentWatermark = newWatermark;
+      return this;
+    }
+
+    /**
+     * Advance the processing time by the specified amount.
+     *
+     * @return this {@link TestStream.Builder}
+     */
+    public Builder<T> advanceProcessingTime(Duration amount) {
+      checkArgument(amount.getMillis() > 0,
+          "Must advance the processing time by a positive amount. Got: ",
+          amount);
+      events.add(ProcessingTimeEvent.<T>advanceBy(amount));
+      return this;
+    }
+
+    /**
+     * Advance the watermark to infinity, completing this {@link TestStream}. Future calls to
+     * the same builder will not affect the returned {@link TestStream}.
+     */
+    public TestStream<T> advanceWatermarkToInfinity() {
+      events.add(WatermarkEvent.<T>advanceTo(BoundedWindow.TIMESTAMP_MAX_VALUE));
+      return new TestStream<>(coder, events.build());
+    }
+  }
+
+
+  /**
+   * An event in a {@link TestStream}. A marker interface for all events that happen while
+   * evaluating a {@link TestStream}.
+   */
+  public interface TestSourceEvent<T> {}
```

Comment:
I would also just call this Event, since someone can import TestStream and refer to this type as TestStream.Event